runtime/common/src/xcm_impl.rs

```diff
@@ -0,0 +1,491 @@
+		AccountId: Clone,
+		ReceiverAccount: Get<AccountId>,
+	> TakeRevenue for XcmFeesToAccount<Assets, Matcher, AccountId, ReceiverAccount>
+{
+	fn take_revenue(revenue: MultiAsset) {
+		match Matcher::matches_fungibles(&revenue) {
+			Ok((asset_id, amount)) =>
+				if !amount.is_zero() {
+					let ok = Assets::mint_into(asset_id, &ReceiverAccount::get(), amount).is_ok();
+					debug_assert!(ok, "`mint_into` cannot generally fail; qed");
+				},
+			Err(_) => log::debug!(
+				target: "xcm",
+				"take revenue failed matching fungible"
+			),
+		}
+	}
+}
+
+pub trait Reserve {
+	/// Returns assets reserve location.
+	fn reserve(&self) -> Option<MultiLocation>;
+}
+
+// Takes the chain part of a MultiAsset
+impl Reserve for MultiAsset {
+	fn reserve(&self) -> Option<MultiLocation> {
+		if let xcmAssetId::Concrete(location) = self.id.clone() {
+			let first_interior = location.first_interior();
+			let parents = location.parent_count();
+			match (parents, first_interior) {
+				// The only case for non-relay chain will be the chain itself.
+				(0, Some(Parachain(id))) => Some(MultiLocation::new(0, X1(Parachain(*id)))),
+				// Only Sibling parachain is recognized.
+				(1, Some(Parachain(id))) => Some(MultiLocation::new(1, X1(Parachain(*id)))),
+				// The Relay chain.
+				(1, _) => Some(MultiLocation::parent()),
+				// No other case is allowed for now.
+				_ => None,
+			}
+		} else {
+			None
+		}
+	}
+}
+
+/// A `FilterAssetLocation` implementation. Filters multi native assets whose
+/// reserve is same with `origin`.
+pub struct MultiNativeAsset;
+
+impl FilterAssetLocation for MultiNativeAsset {
+	fn filter_asset_location(asset: &MultiAsset, origin: &MultiLocation) -> bool {
+		if let Some(ref reserve) = asset.reserve() {
+			if reserve == origin {
+				return true
+			}
+		}
+		false
+	}
+}
+
+/// A set of pallet_config that runtime must implement.
+pub trait BaseRuntimeRequirements:
+	frame_system::Config<BlockNumber = BlockNumber, AccountId = AccountId>
+	+ pallet_balances::Config<Balance = Balance>
+	+ pallet_extrinsic_filter::Config
+	+ pallet_multisig::Config
+	+ parachain_info::Config
+	+ pallet_xcm::Config
+	+ pallet_treasury::Config
+	+ pallet_transaction_payment::Config
+{
+}
+
+pub trait ParaRuntimeRequirements:
+	BaseRuntimeRequirements + pallet_asset_manager::Config<AssetId = AssetId> + pallet_vesting::Config
+{
+}
+
+#[derive(Clone, Eq, Debug, PartialEq, Ord, PartialOrd, Encode, Decode, TypeInfo)]
+pub enum CurrencyId4Compare {
+	SelfReserve,
+	ParachainReserve(Box<MultiLocation>),
+}
+
+// Our currencyId. We distinguish for now between SelfReserve, and Others, defined by their Id.
+#[derive(Clone, Eq, Debug, PartialEq, Encode, Decode, TypeInfo)]
+pub enum CurrencyId<R: BaseRuntimeRequirements> {
+	SelfReserve(PhantomData<R>), // The only parachain native token: LIT
+	ParachainReserve(Box<MultiLocation>), /* Any parachain based asset, including local native minted
+	                              * ones. */
```

Comment:
Can we format this comment a bit please, e.g. use `// ...` in one line

Reply:
done.